File: src/main/java/org/simbrain/network/smile/SmileClassifier.java
package org.simbrain.network.smile;

import org.simbrain.network.connectors.Layer;
import org.simbrain.network.core.Network;
import org.simbrain.util.propertyeditor.EditableObject;
import smile.classification.Classifier;
import smile.classification.SVM;
import smile.math.kernel.PolynomialKernel;
import smile.math.matrix.Matrix;

import java.awt.geom.Rectangle2D;

public class SmileClassifier extends Layer implements EditableObject {

    /**
     * The classifier object.
     */
    private Classifier<double[]> classifier;

    private double[][] trainingInputs = {{0, 0}, {0, 1}, {1, 0}, {1, 1}};

    private int[] targets = {-1, 1, -1, 1};

    /**
     * Construct a neuron array.
     *
     * @param net  parent net
     * @param size number of components in the array
     */
    public SmileClassifier(Network net, int size) {

        var kernel = new PolynomialKernel(2);
        setLabel(net.getIdManager().getProposedId(SmileClassifier.class));

        try {

            var x = new double[][] {{0,0},{0,1},{1,0},{1,1}};
            var y = new int[] {-1,1,1,-1};

            classifier = SVM.fit(x,y, kernel, 1000, 1E-3);
        } catch(Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void update() {
        // TODO
        var result = classifier.predict(getInputs().col(0));
        // setOneHot(result == -1 ? 0 : 1);
    }

    public Classifier<double[]> getClassifier() {
        return classifier;
    }

    public void setClassifier(Classifier<double[]> classifier) {
        this.classifier = classifier;
    }

    public double[][] getTrainingInputs() {
        return trainingInputs;
    }

    public void setTrainingInputs(double[][] trainingInputs) {
        this.trainingInputs = trainingInputs;
    }

    public int[] getTargets() {
        return targets;
    }

    public void setTargets(int[] targets) {
        this.targets = targets;
    }

    @Override
    public String toString() {
        return "SVM classifier " + getLabel();
    }

    @Override
    public Matrix getInputs() {
        return null;
    }

    @Override
    public void addInputs(Matrix inputs) {

    }

    @Override
    public Matrix getOutputs() {
        return null;
    }

    @Override
    public int size() {
        return 0;
    }

    @Override
    public Network getNetwork() {
        return null;
    }

    @Override
    public Rectangle2D getBound() {
        return null;
    }
}

File: src/main/kotlin/org/simbrain/network/smile/SmileClassifier.kt
package org.simbrain.network.smile

import org.simbrain.network.core.Network
import org.simbrain.network.core.activations
import org.simbrain.network.neurongroups.NeuronGroup
import org.simbrain.network.smile.classifiers.SVMClassifier
import org.simbrain.network.subnetworks.Subnetwork
import org.simbrain.network.trainers.ClassificationDataset
import org.simbrain.network.util.Alignment
import org.simbrain.network.util.Direction
import org.simbrain.network.util.alignNetworkModels
import org.simbrain.network.util.offsetNeuronCollections
import org.simbrain.util.UserParameter
import org.simbrain.util.propertyeditor.EditableObject
import org.simbrain.util.propertyeditor.GuiEditable
import org.simbrain.workspace.Producible
import kotlin.reflect.full.primaryConstructor

/**
 * Simbrain wrapper for classifiers at https://haifengl.github.io/classification.html
 *
 * To work in Simbrain, all classifiers must be able to predict a label from a double array input.
 *
 * Outputs are class labels, which are converted into a one-hot output. A number of outputs must be specified, which
 * thus corresponds to the number of class labels.
 *
 * Different classifiers are trained in different ways and thus have different properties and dialogs.
 *
 */
class SmileClassifier(
    val classifier: ClassificationAlgorithm
) : Subnetwork(), EditableObject {

    /**
     * Integer winner produced by the Smile classifier when predict is called.
     */
    var winner = Integer.MIN_VALUE

    /**
     * Returns the label associated with the winning target integer.
     */
    val winningLabel: String?
        @Producible
        get() = classifier.trainingData.labelTargetMap.getInverse(winner) ?: ""

    val inputNeuronGroup = NeuronGroup(classifier.inputSize).apply {
        label = "Input Layer"
        setLayoutBasedOnSize()
    }.also { modelList.add(it) }

    val outputNeuronGroup = NeuronGroup(classifier.outputSize).apply {
        label = "Output Layer"
        setLayoutBasedOnSize()
    }.also { modelList.add(it) }

    init {
        label = classifier.name
        alignNetworkModels(inputNeuronGroup, outputNeuronGroup, Alignment.VERTICAL)
        offsetNeuronCollections(inputNeuronGroup, outputNeuronGroup, Direction.NORTH, 150.0)
        when (classifier.trainingData.labelEncoding) {
            ClassificationDataset.LabelEncoding.Bipolar -> {
                outputNeuronGroup.neuronList[0].label = "Off"
                outputNeuronGroup.neuronList[1].label = "On"
            }
            ClassificationDataset.LabelEncoding.Integer -> {
                classifier.trainingData.labelTargetMap.entries.forEach { (label, index) ->
                    outputNeuronGroup.neuronList.getOrNull(index)?.label = label
                }
            }
        }
    }
    
    /**
     * Train the classifier using the current training data.
     */
    fun train() {
        classifier.apply {
            fit(trainingData.featureVectors, trainingData.getIntegerTargets())
        }
        events.updated.fire()
    }

    context(Network)
    override fun accumulateInputs() {
        inputNeuronGroup.accumulateInputs()
    }

    /**
     * Update the classifier by apply it to inputs and caching the result as output.
     */
    context(Network)
    override fun update() {
        inputNeuronGroup.update()
        if (classifier.model != null) {
            winner = classifier.predict(inputNeuronGroup.activationArray)
            // println("Prediction of ${this.id} is: $winner")
            if (classifier.model != null) {
                outputNeuronGroup.neuronList.activations = try {
                    classifier.getOutputArray(winner)
                } catch (e: IllegalArgumentException) {
                    System.err.println(e.message)
                    DoubleArray(classifier.outputSize) { 0.0 }
                }.toList()
            }
        }
        events.updated.fire()
    }

    override fun toString(): String {
        return "${label} (${classifier.name}): ${classifier.inputSize} -> $outputSize"
    }

    val inputSize get() = classifier.inputSize

    val outputSize get() = classifier.outputSize

    /**
     * Helper class for creating classifiers.
     */
    class ClassifierCreator() : EditableObject {

        @UserParameter(label = "Number of inputs", order = 10)
        var nin = 4

        var nout by GuiEditable(
            initValue = 2,
            label = "Number of outputs (classes)",
            description = "Ignored for some classifiers (e.g. SVM) that can only produce 2 outputs",
            onUpdate = {
                enableWidget(widgetValue(::classifierType) !is SVMClassifier)
            },
            order = 20
        )

        @UserParameter(label = "Classifier Type", showDetails = false, order = 40)
        var classifierType: ClassificationAlgorithm = SVMClassifier()

        override val name = "Classifier"

        fun create(net: Network): SmileClassifier {
            val classifier = classifierType::class.primaryConstructor?.let { constructor ->
                val paramMap = mapOf("inputSize" to nin, "outputSize" to nout)
                val map = constructor.parameters.associateWith { p -> paramMap[p.name] }
                constructor.callBy(map)
            }!!
            return SmileClassifier(classifier)
        }

    }
}